### DISCUSSION TITLE
코드 결과와 이유를 설명해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#81] [클로저]
```js
var funcs = [];

for (var i = 0; i < 3; i++) {
  funcs[i] = function () { return i; }; // ①
}

for (var j = 0; j < funcs.length; j++) {
  console.log(funcs[j]()); // ②
}
```

> 해당코드를 원하는 방향(0,1,2)으로 동작하게 하려면 어떻게 수정해야하나요?

**Comment by publdaze** [comment]
> - 코드 결과와 이유를 설명해주세요
>   ```js
>     var funcs = [];
>     
>     for (var i = 0; i < 3; i++) {
>       funcs[i] = function () { return i; }; // ①
>     }
>     
>     for (var j = 0; j < funcs.length; j++) {
>       console.log(funcs[j]()); // ②
>     }
>     ```
- 첫번째 반복문을 돌면서 funcs 배열에 정의된 함수선언문이 쌓임
- 두번째 반복문을 톨면서 funcs 배열에 차례대로 접근하여 호출함
- 함수의 호출 시점에 함수가 평가되면서 렉시컬 환경을 구성
- 평가 후 함수를 실행할 때 i라는 변수를 참조하므로 식별자 검색 함
- 현재 실행중인 함수의 렉시컬 환경부터 검색 진행
- 존재하지 않으므로 외부렉시컬 환경에 대한 참조에 할당된 상위 렉시컬 환경에서 찾음
- var의 경우 블록 레벨 스코프가 아닌 함수 레벨 스코프이므로 반복 실행되는 코드 블록이 독자적인 렉시컬 환경을 가지지 않으므로
- i 식별자가 최종적으로 검색되는 위치는 전역 렉시컬 환경이 됨
- 따라서 전역변수 i에 할당된 3이라는 값이 반복적으로 3,3,3 출력됨
> - 해당 코드를 원하는 방향(0,1,2)으로 동작하게 하려면 어떻게 수정해야하나요?
- 클로저를 활용하여 즉시 실행함수에 인수로 i를 전달해 독자적 스코프를 가지게 하는 방법
- var가 아닌 let으로 선언하여 블록 레벨 스코프인 let으로 인해 독자적 스코프를 가지도록 하는 방법
- 을 사용하면 0,1,2로 동작 가능

**Comment by pipisebastian** [comment]
```js
var funcs = [];

for (var i = 0; i < 3; i++) {
  funcs[i] = function () { return i; }; // ①
}

for (var j = 0; j < funcs.length; j++) {
  console.log(funcs[j]()); // ②
}
```


- 3,3,3
- 맨 처음, for문을 순회하면서, funcs[i]에 함수 선언문이 할당됩니다. 또한, var키워드로 선언한 i는 전역변수가 되고, 모두 순회를 했을때, i에는 3이 할당됩니다. 이후, 호출할때는 함수에 인자로 i를 넣는게 아닙니다. 그렇기 때문에, return i 에서 참조하는 i를 스코프 체인을 통해 찾는데,  전역변수 i 를 참조하여, 모두 3이 출력됩니다.

> 해당코드를 원하는 방향(0,1,2)으로 동작하게 하려면 어떻게 수정해야하나요?


```js
var funcs = [];

for (var i = 0; i < 3; i++) {
  funcs[i] = (function (index) {
	  return function(){
			 return index;
	  }
  }(i));  
}
```
1. 함수에 i 인자를 주고, 내부함수에서 이를 참조하게 만들어, 내부함수를 외부함수보다 더 오래 유지시킨다면, index 지역변수(매개변수)를 계속 참조할 수 있습니다. 즉, 즉시실행함수에서 독자적인 스코프를 가지기에, index는 계속 유지가되고, index는 i 이므로, 변함없이 0,1,2 출력이 가능해집니다.
2. 혹은 var키워드 대신, let을 사용한다면, 각 반복문마다 렉시컬환경을 만들어, 0,1,2 출력이 가능해집니다.
